### ISSUE TITLE
Level 7_2

### ISSUE THREAD

**Issue by Swagston20** [nus-cs2030/2122-s2#336]
# Summary
The server for 10 is not right

## Description 
I dun know why the server does not rest after serving 9 but rested for previous few? Anyone has this problem?

## Screenshots (if any): 
![Screen Shot 2022-04-15 at 4 21 42 PM](https://user-images.githubusercontent.com/88470259/163576546-9bb14a29-4f08-465f-b9ef-fdbc5101bbde.png)

**Comment by shotnothing** [comment]
Its a probability it might rest.

Main7.java
``` 
        Supplier<Double> restTimes = () ->
            rngRest.nextDouble() < probRest ? genRestPeriod() : 0.0;
```

note the` rngRest.nextDouble() < probRest ? `part

**Comment by jamietan123** [comment]
I think it depends on your implementation. For mine I attached the rest time to each customer rather than server.

**Comment by bryankwe** [comment]
I had the exact same numbers as you. Perhaps our implementation is similar in that we have an additional intermediary class (like TryToServe) between Wait and Serve. 

What I found was that the servers does actually rest, but because our TryToServe class was called before the Done class, the timings were never updated.

I fixed it in a pretty brutal way (I firmly believe there is a more efficient way and I'm interested to know if anyone has other ideas). In Simulate, I went to update the entire PQ to update the TryToServe & Serve events for customers who were queueing under the server who took a rest.

Consequently, my Statistics had some inaccurate calculations as well, and I had to tweak my methods to suit these changes

TLDR; Created a method to update the entire PQ (basically create a new PQ). This method was very inefficient, and I hope that someone else has a better fix so that I can learn as well.

**Comment by mazchew** [comment]
hi @bryankwe sorry, could you explain abit more on your implementation of level7? I think mine is pretty similar and I'm running into some issues as well, thanks!

**Comment by bryankwe** [comment]
Hi, @mazchew 
Once again, my method was super inefficient, but here's the gist of it:

In my Simulate, I had a method to check if it was Done event (Let's call this check 1), if so, then call the get() for the rest time (my rest supplier was solely in Simulate).

In Check 1, if the server actually rests, then I need to update all the other customer's Try / Serve events (eventTime) queueing under that same server. To do this, I created another method to update the entire PQ.

To update the entire PQ, poll each event out from the original PQ and check whether that event is one that you want to update. If it is, then update and add() to your new PQ, else just add() back the same event to your new PQ.

Do share if you have a better way! ATB!

**Comment by mazchew** [comment]
hi @bryankwe, thanks so much for the reply! i think my method is basically identical to yours, but can I ask how you updated the shop to reflect the server as resting? I can get the timings right but customer 3 still goes to server 1 when server1 is still meant to be resting. thanks so much for the help!

**Comment by bryankwe** [comment]
I have a method in Shop `public Shop letServerRest(Server s, double rest){ ... }` which makes the input `server` in the shop rest by adding `rest` to its current `nextAvailTime`. I also have a boolean `isResting`, which I set to `true` if `rest` is != 0

From what you have said, perhaps you are having issues with the assigning of Wait events? After Level 6 I changed my assigning of wait events as such: (**May be a bit overkill and I might have overcomplicated things**... idk)
// If free + resting AND availTime() > eT, return that server
// Elif !free + !resting, return that server
// Elif !free + resting, return that server
// Else return empty

I have attribute `isServing` which is true when there are customers still in queue, or the server is currently serving.
I also have attribute  `isResting`, which is true when servers are resting

I hope I didn't misunderstand ur qn and ATB!
